README.md

```diff
@@ -1,48 +1,162 @@
+# Architecture
+This section is intended as an overview of the codebase architecture for anyone wishing to modify or extend the code for their own purposes. There are 5 main components to the architecture: `Agent`, `Dataset`, `Collator`, `AgentTrainer`, `Evaluator`
 
-Additionally, you can specify whether to give feedback exactly every ``feedback_freq_steps`` for all episodes in the run (``exact``), or on average (using a poisson distribution) every ``feedback_freq_steps`` across the episodes, with this being constant on a per-episode basis (``poisson``).
+### Agent
+The `Agent` class basically represents any offline-RL-ish agent; i.e. something that we want to train from a dataset of sequential environment interactions, and then use as a policy to produce actions given observations. This is what its basic template definition looks like: 
+```python
+@dataclass
+class AgentInput:
+    mission_embeddings: Any
+    states: Any
+    actions: Any
+    rewards: Any
+    returns_to_go: Any
+    timesteps: Any
+    feedback_embeddings: Any
+    attention_mask: Any
 
-### Example
 
-```sh
-python src/_feedback.py --feedback_type "direction" --feedback_mode "simple" --feedback_freq_steps 1 --feedback_freq_type "exact" --env_name "BabyAI-GoToRedBallGrey-v0" --num_episodes 10 --include_timeout False --seed 42
+class Agent(nn.Module):
+    """
+    Agent is the base class used to represent any offline-rl-ish trainable agent.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+
+    def _forward(self, input: AgentInput) -> Any:
+        pass
+
+    def _compute_loss(self, input: AgentInput, output: Any) -> float:
+        pass
+
+    def forward(self, input: AgentInput, **kwargs) -> Dict:
+        output = self._forward(input)
+        loss = self._compute_loss(input, output)
+        return {"loss": loss}
+
+    def get_action(self, input: AgentInput, context=1, one_hot=False):
+        pass
 ```
 
-If the dataset specified doesn't exist yet locally, it will be generated when the ``_feedback.py`` script is run.
+Implementing any particular type of agent/model then mostly consists of overriding the `_forward`, `_compute_loss` and `get_action` methods.
+
+The file `agent/fdt/base.py` contains an `Agent` implementation for the feedback decision transformer. This is the file that contains almost all the relevant logic for the actual FDT model. The files `agent/fdt/atari.py` and `agent/fdt/minigrid.py` contain slightly specialised versions for use with Atari and Minigrid/BabyAI environments respectively. At the time of writing, the only difference is the details of the CNN used to embed image observations.
```

Comment:
we're not actually including any Atari code for this version (remind me to tell you something about this)